File: library/src/main/java/io/github/pelmenstar1/rangecalendar/ColorHelper.java
package io.github.pelmenstar1.rangecalendar;

import android.graphics.Color;

import androidx.annotation.ColorInt;

final class ColorHelper {
    @ColorInt
    public static int withAlpha(@ColorInt int color, int alpha) {
        return (color & 0x00FFFFFF) | (alpha << 24);
    }

    @ColorInt
    public static int makeDarkerColor(@ColorInt int color, float factor) {
        float invFactor = 1f - factor;

        int a = Color.alpha(color);
        int r = Math.round(Color.red(color) * invFactor);
        int g = Math.round(Color.green(color) * invFactor);
        int b = Math.round(Color.blue(color) * invFactor);

        return Color.argb(a,
                Math.min(r, 255),
                Math.min(g, 255),
                Math.min(b, 255)
        );
    }
}

File: library/src/main/java/io/github/pelmenstar1/rangecalendar/ColorHelper.kt
package io.github.pelmenstar1.rangecalendar

import android.graphics.Color
import androidx.annotation.ColorInt
import kotlin.math.roundToInt

@ColorInt
fun Int.withAlpha(alpha: Int): Int {
    return this and 0x00FFFFFF or (alpha shl 24)
}

fun colorLerp(@ColorInt start: Int, @ColorInt end: Int, fraction: Float): Int {
    return Color.argb(
        lerp(Color.alpha(start), Color.alpha(end), fraction),
        lerp(Color.red(start), Color.red(end), fraction),
        lerp(Color.green(start), Color.green(end), fraction),
        lerp(Color.blue(start), Color.blue(end), fraction)
    )
}

@ColorInt
fun Int.darkerColor(factor: Float): Int {
    val invFactor = 1f - factor

    return Color.argb(
        Color.alpha(this),
        (Color.red(this) * invFactor).roundToInt(),
        (Color.green(this) * invFactor).roundToInt(),
        (Color.blue(this) * invFactor).roundToInt(),
    )
}